### DISCUSSION TITLE
Updating the SwiftUI View based on received Midi messages

### DISCUSSION THREAD

**Discussion by SNewtown** [orchetect/swift-midi#184] [Q&A] [answered]
I have the sending and receiving of MIDI messages wired and working based on MIDIKit and I can evaluate the inbound messages with the events listener delegate (which I placed in the ContentView file and referenced accordingly in the AddInput events intialization.  Where I am currently stuck is updating the UI @State variables based on the received MIDI since I can't seem to get the context to work.  I tried different implementations (@StateObject and @Observable) but still can't seem to get the update to work from the listener event since it doesn't have the context of the view.  I have a number of controls to update but passing a reference to each could be done with a struct, I'm just unsure how the listener can have access to update the @State.  If anyone has this working, I would love a sample!

**Comment by SNewtown** [comment]
I finally got "something" to work... using @ObservableObject class to hold the state variable as well as the listener and then I assigned that object to the content view as an @EnvironmentObject... and that "seems" to be working.  If there is a better way to do this, I'm all ears!

**Comment by orchetect** [comment]
Receiving MIDI is an asynchronous background callback. In order to loop it into SwiftUI's state updates you would need a class that conforms to `ObservableObject`, and you would need a variable marked `@Published` within it. That class needs to be instanced within a SwiftUI View and its declaration should be marked `@ObservedObject`. Don't use `@StateObject` for this.

When receiving MIDI events, update the published var. Then you can point to that variable from a SwiftUI view and it will update from changes.

Passing the object through the view hierarchy with `@EnvironmentObject` is fine as long as its origin declaration is `@ObservedObject`.

This is all basic Combine / SwiftUI stuff, and isn't particular to MIDIKit's operation.

This is the gist of it below. But since `MIDIEvent` doesn't conform to `Identifiable` (and won't ever), in a `List` or `ForEach` we need to either use the array index or a wrap `MIDIEvent` in a custom type that does conform to `Identifiable`.

```swift
@main struct MyApp: App {
    @ObservedObject var midiHelper = MIDIHelper()

    var body: some Scene {
        WindowGroup {
            ContentView()
                .environmentObject(midiHelper)
        }
    }
}

struct ContentView: View {
    @EnvironmentObject var midiHelper: MIDIHelper
    
    var body: some View {
        List(midiHelper.receivedEvents.indices, id: \.self) { index in
            Text(midiHelper.receivedEvents[index].description)
        }
    }
}

class MIDIHelper: ObservableObject {
    @Published var receivedEvents: [MIDIEvent] = []
}
```

**Reply by orchetect** [reply, marked as answer]
Actually if you check out the [EndpointPickers](https://github.com/orchetect/MIDIKit/tree/main/Examples/SwiftUI%20Multiplatform/EndpointPickers) example project it uses this paradigm.

**Reply by SNewtown** [reply]
Yes, that looks a LOT like where I ended up... VERY MUCH appreciated, at the very least it confirms best practice.... even though I eventually got it to work, I was still questioning if this was the "proper" way to do it... so again, thank you!

And I will definitely check out the EndpointPickers example!

Cheers!

**Reply by orchetect** [reply]
Yeah and to be fair it's not super intuitive which is why the examples are there. When SwiftUI is in the mix, you really need a helper class approach.
